packages-exp/auth-exp/src/core/strategies/credential.ts

```diff
@@ -39,3 +46,52 @@ export async function signInWithCredential(
   await auth.updateCurrentUser(userCredential.user);
   return userCredential;
 }
+
+export async function linkWithCredential(
+  userExtern: externs.User,
+  credentialExtern: externs.AuthCredential
+): Promise<UserCredential> {
+  const user = userExtern as User;
+  const credential = credentialExtern as AuthCredential;
+  await _assertLinkedStatus(false, user, credential.providerId);
+
+  const response = await credential._linkToIdToken(
+    user.auth,
+    await user.getIdToken()
+  );
+
+  const newCred = _authCredentialFromTokenResponse(response);
+  await user._updateTokensIfNecessary(response, /* reload */ true);
+  return new UserCredentialImpl(user, newCred, OperationType.LINK);
+}
+
+export function _authCredentialFromTokenResponse(
+  response: PhoneOrOauthTokenResponse
+): AuthCredential | null {
+  const {
+    temporaryProof,
+    phoneNumber
+  } = response as SignInWithPhoneNumberResponse;
+  if (temporaryProof && phoneNumber) {
+    return new PhoneAuthCredential({ temporaryProof, phoneNumber });
+  }
+
+  // TODO: Handle Oauth cases
+  return null;
+}
+
+export async function _assertLinkedStatus(
+  expected: boolean,
+  user: User,
+  provider: externs.ProviderId
+): Promise<void> {
+  await _reloadWithoutSaving(user);
```

Comment:
there's a potential race condition if this gets called twice, do we care?

Reply:
Discussed offline, it should be fine